Update change note history

alex-ju opened the pull request:
This PR:

- Updates the change notes layout to use a  2/3 column for the internal notes timeline
- Make internal notes permanently visible
- Replaces the accordion with a [timeline entry pattern used consistently with content-publisher](https://github.com/alphagov/content-publisher/blob/60fa3c2fd2b351b2ca3ad5d1bd39f3ea1527918a/app/views/documents/show/_history_tab.html.erb)
- Use the accordion at the edition level (rather then internal note entry level).

<table>
<tr><th>Before</th><th>After</th></tr>
<tr><td valign="top">

![internal-notes-before](https://user-images.githubusercontent.com/788096/63172172-ff907380-c034-11e9-89a7-c1aef56f1376.png)

</td><td valign="top">

![collections-publisher dev gov uk_step-by-step-pages_1_internal-change-notes copy](https://user-images.githubusercontent.com/788096/63273522-0964e180-c296-11e9-98aa-ccf5727eb3e1.png)

</td></tr>
</table>

**Note**: I appreciate the history of this PR doesn't make much sense as there was a bit of a back and forth on it, so I'd suggesting file-based review and I can squash them before merging.

[Trello card](https://trello.com/c/pmUvssRI)

leenagupte commented on app/views/step_by_step_pages/internal_change_notes.html.erb:
```diff
@@ -25,56 +25,46 @@
-          },
-          expanded: index == 0
+      items: @step_by_step_page.internal_change_notes
+        .group_by { |edition| edition.edition_number }
```
I think it's fine to use `group_by` here 👍

leenagupte commented on app/views/step_by_step_pages/internal_change_notes.html.erb:
```diff
@@ -25,56 +25,46 @@
-          expanded: index == 0
+      items: @step_by_step_page.internal_change_notes
+        .group_by { |edition| edition.edition_number }
+        .map { |edition_number, records|
```
I don't think the `edition_number` being defined here is being used anywhere.

alex-ju replied:
`group_by` returns a hash where the key is the `edition_number` which is why we need to reflect both arguments on the `map` method.

leenagupte replied:
You can use `_` to ignore a return value if it's not used.

alex-ju replied:
Nice! Updated.

leenagupte commented on app/views/step_by_step_pages/internal_change_notes.html.erb:
```diff
@@ -25,56 +25,46 @@
+                    <p class='app-c-timeline-entry__dateline'>#{note.readable_created_date} by #{note.author}</p>
+                    <pre class='app-c-timeline-entry__description'>#{note.description}</pre>
+                  </div>"
+                }.join()
```
What is the `join` doing?

alex-ju replied:
we're using `join` to concatenate/merge an array of strings into a single string to be passed as content for the accordion item. `flat_map` would flatten the array but we'll still need to merge the elements.

leenagupte commented on app/views/step_by_step_pages/internal_change_notes.html.erb:
```diff
@@ -25,56 +25,46 @@
+                }.join()
+              )
+            },
+            expanded: !records.first.edition_number
```
This reads a bit strangely, but it seems to work.

leenagupte commented on spec/features/managing_steps_spec.rb:
```diff
@@ -203,12 +203,12 @@ def and_I_see_the_steps_updated_on_the_step_by_step_details_page
 
   def and_I_write_a_change_note
-    fill_in "Description", with: "I've changed this step by step!"
+    fill_in "Internal note", with: "I've changed this step by step!"
```
It feels like we're missing tests about how change notes are displayed.

alex-ju replied:
Added tests to capture the new internal notes' structure on `then_the_change_note_should_be_saved`. After we check for the success message we're now also checking that the history timeline was updated and contains the newly inserted note.

alex-ju commented:
Thanks for the review @leenagupte. Tried to address the comments.

> When you sort out the commits, could you add a reason for the changes? e.g. `bindCancelAddChangeNoteLink` isn't needed because the new change note text area is always being displayed now etc.

Added more details on 'why' for each commit message.

> At the moment we don't record state changes to step by steps in same way that content-publisher does, but I wonder if in a future enhancement we could change InternalChangeNote to accept a "change_type" and use that to populate the heading, for example, "Draft saved", "Published" and "Internal note" could be the default heading for hand craft notes.

Yes, having a `change_type` sounds like a great idea. This is the reason I kept the type in the title - even though at the moment there's only one.

leenagupte commented on app/views/step_by_step_pages/internal_change_notes.html.erb:
```diff
@@ -24,26 +24,30 @@
-    } %>
+    <% @step_by_step_page.internal_change_notes.each do |entry| %>
+      <div class="app-c-timeline-entry">
+        <% if displayed_edition_number != entry.edition_number %>
```
What happens here if `entry.edition_number` is nil? Originally it displayed `Current version`.

Is the check needed? Will `displayed_edition_number` ever be the same as `entry.edition_number` at this point in the loop?

alex-ju replied:
Not anymore, I've removed it in one of the following commits when we changed the approach.

leenagupte replied:
Could you remove it from the history?

alex-ju replied:
It is needed in this commit and it's doesn't always evaluate to true. It [starts with nil](https://github.com/alphagov/collections-publisher/pull/694/commits/c3105a241af4d66e663f6bc23c33c14c8d74abff#diff-5a189794e316cb715d3507edf2b23b8fR28) then [gets updated every time we reach a new edition](https://github.com/alphagov/collections-publisher/pull/694/commits/c3105a241af4d66e663f6bc23c33c14c8d74abff#diff-5a189794e316cb715d3507edf2b23b8fR52). We're using [a similar technique in content-publisher](https://github.com/alphagov/content-publisher/blob/master/app/views/documents/show/_history_tab.html.erb#L7).

leenagupte replied:
I don't get how this working, but I see that it does hide the title, there by "grouping" the change notes. so 👍

leenagupte commented on app/views/step_by_step_pages/internal_change_notes.html.erb:
```diff
@@ -24,26 +24,30 @@
+    <% @step_by_step_page.internal_change_notes.each do |entry| %>
+      <div class="app-c-timeline-entry">
+        <% if displayed_edition_number != entry.edition_number %>
+          <h3 class="app-c-timeline-entry__title"><%= "#{entry.edition_number.to_i.ordinalize} edition" %></h3>
```
What happens here if `entry.edition_number` is nil? Originally it displayed `Current version`.

alex-ju replied:
This is also not present in the final changes. It looks like this after the last commit `"#{records.first.edition_number ? "#{records.first.edition_number.to_i.ordinalize} edition" : "Current version"}"`

leenagupte replied:
To answer the original question, at this point in the code, the history looks like:

![Screen Shot 2019-08-21 at 11 43 10](https://user-images.githubusercontent.com/5793815/63426413-a6488b80-c40a-11e9-96a7-bbc3821d32fb.png)

Which is fine, but it's not clear that the first title is being lost for step by steps with a draft from the commit description.

leenagupte commented on app/views/step_by_step_pages/internal_change_notes.html.erb:
```diff
@@ -28,6 +28,23 @@
 <% displayed_edition_number = nil %>
 <div class="govuk-grid-row">
   <div class="govuk-grid-column-two-thirds">
+    <%= form_for(@internal_change_note, url: step_by_step_page_internal_change_notes_path) do |form| %>
```
Can this change be fixed the commit [Move internal notes to aside 1/3"](3dfb1c621d2613b0c12fde25b3b641d39f453485)

alex-ju replied:
Yep, will merge a few commits as we pivoted on the approach.

leenagupte commented on app/views/step_by_step_pages/internal_change_notes.html.erb:
```diff
@@ -45,25 +44,27 @@
+                    <p class='app-c-timeline-entry__dateline'>#{note.readable_created_date} by #{note.author}</p>
+                    <pre class='app-c-timeline-entry__description'>#{note.description}</pre>
+                  </div>"
+                }.join()
```
Nitpick: In Ruby you don't need the empty parentheses. You can just call `.join`.

alex-ju commented:
Thanks for the review @leenagupte. Cleaned the commit history from the pivoting approach. It's only 3 clear commits now. Updated with your suggestions too.

leenagupte commented on app/views/step_by_step_pages/internal_change_notes.html.erb:
```diff
@@ -45,25 +44,27 @@
 
-    <% @step_by_step_page.internal_change_notes.each do |entry| %>
-      <div class="app-c-timeline-entry">
-        <% if displayed_edition_number != entry.edition_number %>
```
As this check always evaluated to true, and therefore was never needed, can you fix up it's removal to the commit it was added in.

leenagupte commented on app/views/step_by_step_pages/internal_change_notes.html.erb:
```diff
@@ -25,56 +25,45 @@
+<% displayed_edition_number = nil %>
 <div class="govuk-grid-row">
   <div class="govuk-grid-column-two-thirds">
-    <%= render "govuk_publishing_components/components/accordion", {
```
Could you add a line to the commit message that clarifies that the accordion is being removed to make it easier to change the styling, but that it will be re-added in the next commit? I assume that's the reason the accordion is being removed here.

alex-ju commented:
@leenagupte updated commit messages as requested
